### DISCUSSION TITLE
functions as compilable predicates

### DISCUSSION THREAD

**Discussion by jedi1992** [EvgSkv/logica#112] [Q&A] [answered]
Hi,
In the GetPhone, GetEmail example https://github.com/EvgSkv/logica/blob/main/examples/scripts/outer_join.l

a) GetEmail(name) = x :-  Email(name:,  email:x);

can also be written as 

b) GetEmail(name) = email :-
  email AnyValue= (x :- Email(name:, email: x));

If I use definition a) I can compile it (%%logica GetEmail) and use the function as a look up as well as "materialize it". 
If I use definition b) then I can use it only as a lookup  function.

Why 2 ways - any significance ? 
Is there some under the hood idea  -  function as compilable predicates  and converting a lookup to table loookup   beforehand  or something ( performance benefits thereof). 

Thanks.

**Comment by EvgSkv** [comment, marked as answer]
Hello!

The fact that a) can be rendered and b) can't is irrelevant to them being written as functions. It would work the same way if they were regular predicates:

```
GetEmailA(name:, email:) :- Email(name:, email:);
GetEmailB(name:, email:) :-
  email AnyValue= (x :- Email(name:, email: x));
```

These two ways are defining different mathematical predicates and are used for different purposes.
Predicate a) is *finite*, it has as many rows as original `Email` table.
Predicate b) is mathematically *infinite* it has as many rows as there are names, for each name that's in the original Email table having one of emails of the person and for each name that's not in the original table having `null` value.

Another difference is what you pointed out: predicate a) is `concrete`, i.e. it can be computed as a table and predicate b) is not concrete, it can only be called from other predicates.

Predicate b) is not concrete because there is no way to assign `name` variable. It's used in the body inside of aggregation and usage inside of aggregation doesn't assign it a value, it uses the value that has to be assigned by the caller of the predicate. This is important, as it actually lets us define the predicate to be infinite, producing a row for any input name, even if it's not in the original table.

Predicate b) is actually doing what's called `outer join` in SQL.

This is a subtle subject. Let me know if you have further questions.
